File: android/src/main/java/com/iologinutils/CustomTabActivityHelper.java
package com.iologinutils;

import android.app.Activity;
import android.content.ComponentName;
import android.net.Uri;
import android.util.Log;

import androidx.browser.customtabs.CustomTabsClient;
import androidx.browser.customtabs.CustomTabsIntent;
import androidx.browser.customtabs.CustomTabsServiceConnection;
import androidx.browser.customtabs.CustomTabsSession;

import com.facebook.react.bridge.Promise;

public class CustomTabActivityHelper extends CustomTabsServiceConnection{
  private CustomTabsSession mCustomTabsSession;
  public static CustomTabsClient mClient;
  private CustomTabsServiceConnection mConnection;
  public static Object lock = new Object();


  public static void openCustomTab(Activity activity,
                                   CustomTabsIntent customTabsIntent,
                                   Uri uri,
                                   Promise promise) {
    String packageName = CustomTabsHelper.getPackageNameToUse(activity,uri);

    if (packageName == null) {
      promise.reject("error", "missing browser");
    } else {
      customTabsIntent.intent.setPackage(packageName);
      customTabsIntent.launchUrl(activity, uri);
    }
  }

  public void unbindCustomTabsService(Activity activity) {
    if (mConnection == null) return;
    activity.unbindService(mConnection);
    mClient = null;
    mCustomTabsSession = null;
    mConnection = null;
  }

  public CustomTabsSession getSession() {
    if (mClient == null) {
      mCustomTabsSession = null;
    } else if (mCustomTabsSession == null) {
      mCustomTabsSession = mClient.newSession(null);
    }
    return mCustomTabsSession;
  }

  public void bindCustomTabsService(Activity activity,Uri uri) {
    if (mClient != null) return;

    String packageName = CustomTabsHelper.getPackageNameToUse(activity,uri);
    if (packageName == null) return;

    mConnection = this;
    CustomTabsClient.bindCustomTabsService(activity, packageName, mConnection);
  }

  public boolean mayLaunchUrl(Uri uri) {
    if (mClient == null) return false;

    CustomTabsSession session = getSession();
    if (session == null) return false;

    return session.mayLaunchUrl(uri, null, null);
  }

  @Override
  public void onCustomTabsServiceConnected(ComponentName name, CustomTabsClient client) {
    synchronized(lock) {
      mClient = client;
      mClient.warmup(0L);
      lock.notifyAll();
    }
  }

  @Override
  public void onServiceDisconnected(ComponentName name) {
    mClient = null;
    mCustomTabsSession = null;
  }
}

File: android/src/main/java/com/iologinutils/CustomTabActivityHelper.kt
package com.iologinutils

import android.app.Activity
import android.content.ComponentName
import android.content.Intent
import android.net.Uri
import androidx.browser.customtabs.CustomTabsClient
import androidx.browser.customtabs.CustomTabsIntent
import androidx.browser.customtabs.CustomTabsServiceConnection
import androidx.browser.customtabs.CustomTabsSession
import com.facebook.react.bridge.Promise
import com.iologinutils.CustomTabsHelper.getPackageNameToUse
import com.iologinutils.IoLoginUtilsModule.Companion.generateErrorObject
import okhttp3.internal.notifyAll

class CustomTabActivityHelper : CustomTabsServiceConnection() {
  private var mCustomTabsSession: CustomTabsSession? = null
  private var mConnection: CustomTabsServiceConnection? = null
  fun unbindCustomTabsService(activity: Activity) {
    mConnection?.let {
      activity.unbindService(it)
    }
    mClient = null
    mCustomTabsSession = null
    mConnection = null
  }

  private val session: CustomTabsSession?
    get() {
      mClient?.let {
        mCustomTabsSession = mCustomTabsSession ?: it.newSession(null)
      }
      return mCustomTabsSession
    }

  fun bindCustomTabsService(activity: Activity?, uri: Uri?) {
    mClient?.let { return }
    activity?.let { validActivity ->
      val packageName = getPackageNameToUse(validActivity, uri) ?: return
      mConnection = this
      CustomTabsClient.bindCustomTabsService(validActivity, packageName, this)
    }
  }

  fun mayLaunchUrl(uri: Uri?): Boolean = mClient?.let { session?.mayLaunchUrl(uri, null, null) } ?: false

  override fun onCustomTabsServiceConnected(name: ComponentName, client: CustomTabsClient) {
    synchronized(lock) {
      mClient = client
      client.warmup(0L)
      lock.notifyAll()
    }
  }

  override fun onServiceDisconnected(name: ComponentName) {
    mClient = null
    mCustomTabsSession = null
  }

  companion object {
    var mClient: CustomTabsClient? = null
    var lock = Any()
    fun openCustomTab(
      activity: Activity,
      customTabsIntent: CustomTabsIntent,
      uri: Uri,
      promise: Promise
    ) {
      val packageName = getPackageNameToUse(activity, uri)
      if (packageName == null) {
        promise.reject("NativeAuthSessionError", generateErrorObject("MissingBrowserPackageNameWhileOpening", null, null, null))
        return
      } else {
        customTabsIntent.intent.setPackage(packageName)
        customTabsIntent.intent.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY)
        customTabsIntent.intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        customTabsIntent.intent.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
        customTabsIntent.intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        customTabsIntent.launchUrl(activity, uri)
      }
    }
  }
}
